### PR TITLE
Ask user before allowing kdbx files as key files

### PR DESCRIPTION
**Pull request by qw3ry**
## Type of change
- ✅ New feature (non-breaking change which adds functionality)

## Description and Context
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Describe the context of your change. Explain large code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" as necessary )
When adding a keyfile, the selected file is checked:
- if it is a `*.kdbx` file, the user is warned that the file is not appropriate, but can ignore the warning. Maybe some users decided to give their keyfiles a kbdx extension and want to continue with this in the future.
- if it is the current database file, an error is shown. The database would be corrupted if it was used.

Strictly speaking, the change could be considered to be a breaking change. The user cannot pick the database as its own keyfile anymore. But practically this would corrupt the database anyway, so no ones "workflow" is breaking

This change might need new translations (I'm not sure how they work with Qt. Tested with English local only).

In issue #3611 @phoerious and @droidmonkey kind of referenced this feature

## Screenshots
[TIP]:  # ( Do not include screenshots of your actual database! )
Sorry for the missing window headers, my screenshot tool failed to grab them. They are (in both cases): "Database file as key file"
image

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
Only manual test. This is a GUI change, there is not much to test here.

## Checklist:
[NOTE]: # ( Please go over all the following points. )
[NOTE]: # ( Again, remove any lines which don't apply. )
[NOTE]: # ( Pull Requests that don't fulfill all [REQUIRED] requisites are likely )
[NOTE]: # ( to be sent back to you for correction or will be rejected.  )
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

### CODE REVIEW

**Comment by droidmonkey** [comment]
This could be more generic in that any file chosen that is not a *.key file should provide a warning about how a key file cannot change or will invalidate your database access. Great start

**Comment by phoerious** [comment]
I don't think we should enforce any specific file extension. People want to use their jpegs and stuff as key files.

**Comment by qw3ry** [comment]
@phoerious Thats what I thought as well. The file picker _does_ have a *.key filter as pre-selection already, that should suffice (?)

**Comment by droidmonkey** [comment]
My comment still stands, if there user actively chooses to go "off filter" they should be aware of the consequences of that choice. A prompt may not be entirely appropriate, but a warning banner in the master key tool could work well.

**Comment by qw3ry** [comment]
@droidmonkey not sure how to implement this in a reasonable way:

 - The key-edit dialog seems to be `MessageBox`-centered at first sight. Is there a `MessageWidget` anywhere near, that I can easily access?
- The `MainWindow` does have `displayGlobalMessage` that targets a `MessageWidget`, but the method is not static, so I'd need to carry an object pointer through several levels of constructors - or "hope" that the parent pointer in `QWidget` is of correct type.
- I could add another message widget inside the settings, but I'm not sure if thats a good idea either.

**Comment by droidmonkey** [comment]
MessageWidget is overkill for this. A simple label with a note in it is perfectly fine. It could even be static just reminding users not to pick their database file or any file that is prone to changing. Prefer a freshly generated key file.

File: src/gui/masterkey/KeyFileEditWidget.ui [hunk 1]
```diff
@@ -63,6 +63,13 @@
      </property>
     </spacer>
    </item>
+   <item row="1" column="0">
+    <widget class="QLabel" name="label">
+     <property name="text">
+      <string>Do not use a file that may change! Your database would be lost forever.</string>
+     </property>
+    </widget>
+   </item>
```
Comment (qw3ry) [inline src/gui/masterkey/KeyFileEditWidget.ui]: @droidmonkey like this?

Comment (qw3ry) [reply src/gui/masterkey/KeyFileEditWidget.ui]: The label text feels a bit strange to me, but Im not a native speaker, and I don't have a better idea. Feel free to suggest something

**Comment by droidmonkey** [comment]
I made the necessary changes

File: src/gui/masterkey/KeyFileEditWidget.cpp [hunk 2]
```diff
@@ -120,6 +122,25 @@ void KeyFileEditWidget::browseKeyFile()
...
+                                "file or generate a new key file."));
+        return;
+    } else if (fileName.endsWith(".kdbx", Qt::CaseInsensitive)) {
+        auto response =
+            MessageBox::warning(getMainWindow(),
+                                tr("Invalid Key File"),
+                                tr("You have chosen a database file as your key file. This file will most likely "
+                                   "change and prevent you from unlocking your database. Do you want to continue?"),
+                                MessageBox::Continue | MessageBox::Cancel,
+                                MessageBox::Cancel);
```
Comment (phoerious) [inline src/gui/masterkey/KeyFileEditWidget.cpp]: Please rephrase to something a bit more concise, such as:

Title: "Suspicious key file"
Message: "The chosen key file looks like a password database file.
A key file must be a static file that never changes or you will lose access to your database forever.
Are you sure you want to continue with this file?"

Comment (droidmonkey) [reply src/gui/masterkey/KeyFileEditWidget.cpp]: Changes made

**Pull request merged**